### ISSUE TITLE
With tablets, views are not finished building when interrupted by decommission

### ISSUE THREAD

**Issue by nyh** [scylladb/scylladb#17293]
I have seen this problem in several MV-related dtests that fail with tablets enabled (`--tablets`):
* `materialized_views_test.py::TestMaterializedViews::test_mv_populating_from_existing_data_during_node_decommission`
* `materialized_views_test.py::TestMaterializedViews::test_mv_resurrected_rows_after_decommission_interrupt` (see also #17027, which we may later close as a duplicate)

The first test is easiest to understand: The test starts a 4-node cluster, creates a base table with data and starts building a view on it, and in parallel **decomissions** the second node. The problem is that the decomissioned node never shuts down, which confuses the test (it sees that node alive, and then expects it to have done view building, which of course it hasn't - so it declares that the test failed because view building hasn't finished).

The logs on the second node shows that the decomission was done:
```
INFO  2024-02-12 17:30:01,390 [shard 0:strm] storage_service - DECOMMISSIONING: done
INFO  2024-02-12 17:30:02,115 [shard 0:main] raft_group_registry - marking Raft server 9512a30d-efd3-4d4a-ae03-51f0ca135e6d as dead for raft groups
INFO  2024-02-12 17:30:02,120 [shard 0:main] raft_group_registry - marking Raft server 7067953c-a1b3-4657-945d-2ddc00309ad2 as dead for raft groups
INFO  2024-02-12 17:30:02,126 [shard 0:main] raft_group_registry - marking Raft server be787437-5fed-4088-83ef-9b0c5e646df6 as dead for raft groups
```

but then, the node stays alive:

90 seconds after the decomission is done the log reports
```
WARN  2024-02-12 17:31:33,435 [shard 0:main] service_level_controller - update_from_distributed_data: failed to update configuration for more than  90 seconds : exceptions::read_timeout_exception (Operation timed out for system_distributed.service_levels - received only 0 responses from 1 CL=ONE.)
```
and then the same message again and again with increasing timeouts:
```
WARN  2024-02-12 17:32:53,436 [shard 0:main] service_level_controller - update_from_distributed_data: failed to update configuration for more than  180 seconds : exceptions::read_timeout_exception (Operation timed out for system_distributed.service_levels - received only 0 responses from 1 CL=ONE.)
```
```
WARN  2024-02-12 17:34:33,439 [shard 0:main] service_level_controller - update_from_distributed_data: failed to update configuration for more than  270 seconds : exceptions::read_timeout_exception (Operation timed out for system_distributed.service_levels - received only 0 responses from 1 CL=ONE.)
```
and so on.

This loop never ends until dtest declares the test a failure and kills Scylla gently - interestingly, at that point, the shutdown proceeds successfully.

**Comment by bhalevy** [comment]
@xemul please look into this issue

**Comment by nyh** [comment]
I wonder if this is a duplicate of the older https://github.com/scylladb/scylladb/issues/13925, which reports a similar problem with just raft topology changes enabled, without tablets.

**Comment by xemul** [comment]
First observation -- test does

- decommission node2
- wait for views to be built
- stop node2

With tablets it hands in step 2 -- waiting for views -- so node2 staying alive up until test times out is not a bug. Decommision with tablets finishes and test notices it (with the respective message in its logs)

**Comment by xemul** [comment]
Regarding step 2

Logs for vnodes case (formatted for ease of reading)

```
tables_view_manager.py:675  | test_mv_populating_from_existing_data_during_node_decommission: Waiting for view ks.tm_table_mv_1 to finish building: num_nodes=5 num_alive=5 cl=QUORUM
tables_view_manager.py:684  | 
  test_mv_populating_from_existing_data_during_node_decommission:
    wait_for_view ks.tm_table_mv_1:
      status=defaultdict(<class 'set'>, {'SUCCESS': {
    UUID('8fd7ad94-4e78-474e-80ed-203e6dd74fc9'),
    UUID('293ba7fb-16f3-46fa-a89a-8fde92eb0e24'),
    UUID('391f6c73-f0e8-42af-9ba0-e861b1252c42'),
    UUID('93995b4b-f385-49fd-9fc0-9576af6d459f'),
    UUID('040ca372-d308-4514-9b68-e80a3c86c6ac')}})
```

logs for tablets case (formatted as well)

```
tables_view_manager.py:675  | test_mv_populating_from_existing_data_during_node_decommission: Waiting for view ks.tm_table_mv_1 to finish building: num_nodes=5 num_alive=5 cl=QU
ORUM
tables_view_manager.py:684  |
  test_mv_populating_from_existing_data_during_node_decommission:
    wait_for_view ks.tm_table_mv_1:
      status=defaultdict(<class 'set'>, {'SUCCESS': {
    UUID('7fb379af-7d50-4900-bbce-68c41b52cbb8'),
    UUID('cf8ebd68-b50c-44f5-8cea-5400d3da624b'),
    UUID('6cc76b74-a6dc-410a-a87d-981d88cb8644'),
    UUID('4c9765a8-0966-4105-8371-dbcd52835f8e')},
 'STARTED': {UUID('f8c36292-24b8-4bcf-968e-e05206e3db97')}
})
```

So one of the ... nodes (?) didn't have its view status updated to SUCCESS

**Comment by xemul** [comment]
The waiter executes `"SELECT host_id, status FROM system_distributed.view_build_status"` query and waits for all the records to have `SUCCESS` status, otherwise it waits for 5 seconds and repeats

**Comment by xemul** [comment]
The node that fails to upgrade its status to SUCCESS (tablets case) is 
```
node2.log:INFO  2024-02-26 16:32:43,581 [shard 0:main] init - Setting local host id to f8c36292-24b8-4bcf-968e-e05206e3db97
```
the node that is being decommissioned

**Comment by xemul** [comment]
@nyh , I'm not familiar with MVs, so before I find the answer myself you might have it already -- when decommissioning a node in _non_-tablets case, where does it finishes its view building and marks it with SUCCESS status?

**Comment by xemul** [comment]
From node2 logs

No-tablets case
```
INFO  2024-02-26 16:29:47,703 [shard 0:strm] view - Building view ks.tm_table_mv_0, starting at token minimum token
INFO  2024-02-26 16:29:47,767 [shard 0:strm] view - Building view ks.tm_table_mv_1, starting at token -7658574230999992139
INFO  2024-02-26 16:29:47,795 [shard 0:strm] view - Building view ks.tm_table_mv_2, starting at token -7339353085380412756
INFO  2024-02-26 16:29:47,864 [shard 0:strm] view - Building view ks.tm_table_mv_3, starting at token -6200736432500661424
INFO  2024-02-26 16:29:47,901 [shard 0:strm] view - Building view ks.tm_table_mv_4, starting at token -5591888732542468378
INFO  2024-02-26 16:29:47,971 [shard 0:strm] view - Building view ks.tm_table_mv_5, starting at token -4736430567669200502
INFO  2024-02-26 16:29:48,004 [shard 0:strm] view - Building view ks.tm_table_mv_6, starting at token -4249517068916946426
INFO  2024-02-26 16:29:48,085 [shard 0:strm] view - Building view ks.tm_table_mv_7, starting at token -3525099408058616206
INFO  2024-02-26 16:29:48,106 [shard 0:strm] view - Building view ks.tm_table_mv_8, starting at token -3412146071066557249
INFO  2024-02-26 16:29:48,194 [shard 0:strm] view - Building view ks.tm_table_mv_9, starting at token -2858281918006969432
INFO  2024-02-26 16:29:49,959 [shard 0:strm] view - Finished building view ks.tm_table_mv_0
INFO  2024-02-26 16:29:50,193 [shard 0:strm] view - Finished building view ks.tm_table_mv_1
INFO  2024-02-26 16:29:50,232 [shard 0:strm] view - Finished building view ks.tm_table_mv_2
INFO  2024-02-26 16:29:50,401 [shard 0:strm] view - Finished building view ks.tm_table_mv_3
INFO  2024-02-26 16:29:50,448 [shard 0:strm] view - Finished building view ks.tm_table_mv_4
INFO  2024-02-26 16:29:50,528 [shard 0:strm] view - Finished building view ks.tm_table_mv_5
INFO  2024-02-26 16:29:50,553 [shard 0:strm] view - Finished building view ks.tm_table_mv_6
INFO  2024-02-26 16:29:50,565 [shard 0:strm] api - decommission
INFO  2024-02-26 16:29:50,741 [shard 0:strm] view - Finished building view ks.tm_table_mv_7
INFO  2024-02-26 16:29:50,780 [shard 0:strm] view - Finished building view ks.tm_table_mv_8
INFO  2024-02-26 16:29:50,943 [shard 0:strm] view - Finished building view ks.tm_table_mv_9
INFO  2024-02-26 16:31:51,642 [shard 0:main] view - Draining view builder
INFO  2024-02-26 16:31:51,642 [shard 1:main] view - Draining view builder
INFO  2024-02-26 16:31:52,047 [shard 0:main] view - Stopping view builder
INFO  2024-02-26 16:31:52,047 [shard 1:main] view - Stopping view builder
```

it finishes most of views building _before_ decommission starts, then finishes the remaining. OK, I guess that's expected

In tablets case it's much worse

```
INFO  2024-02-26 16:33:21,275 [shard 0:strm] view - Building view ks.tm_table_mv_0, starting at token minimum token
INFO  2024-02-26 16:33:21,357 [shard 0:strm] view - Building view ks.tm_table_mv_1, starting at token 826723043570522844
INFO  2024-02-26 16:33:21,497 [shard 0:strm] view - Building view ks.tm_table_mv_2, starting at token 891310248864169186
INFO  2024-02-26 16:33:21,499 [shard 0:strm] view - Building view ks.tm_table_mv_3, starting at token 891310248864169186
INFO  2024-02-26 16:33:21,501 [shard 0:strm] view - Building view ks.tm_table_mv_4, starting at token 891310248864169186
INFO  2024-02-26 16:33:21,502 [shard 0:strm] view - Building view ks.tm_table_mv_5, starting at token 891310248864169186
INFO  2024-02-26 16:33:21,502 [shard 0:strm] view - Building view ks.tm_table_mv_6, starting at token 891310248864169186
INFO  2024-02-26 16:33:21,611 [shard 0:strm] view - Building view ks.tm_table_mv_7, starting at token 955142175900778860
INFO  2024-02-26 16:33:21,611 [shard 0:strm] view - Building view ks.tm_table_mv_8, starting at token 955142175900778860
INFO  2024-02-26 16:33:21,612 [shard 0:strm] view - Building view ks.tm_table_mv_9, starting at token 955142175900778860
INFO  2024-02-26 16:33:24,173 [shard 0:strm] api - decommission
INFO  2024-02-26 16:33:30,538 [shard 0:strm] view - Finished building view ks.tm_table_mv_0
INFO  2024-02-26 16:37:23,322 [shard 0:main] view - Draining view builder
INFO  2024-02-26 16:37:23,323 [shard 1:main] view - Draining view builder
INFO  2024-02-26 16:37:23,448 [shard 0:main] view - Stopping view builder
INFO  2024-02-26 16:37:23,448 [shard 1:main] view - Stopping view builder
```
Finishes only one

**Comment by xemul** [comment]
(but I interrupted the --tablets run before it finished, let me re-run it for longer time)

**Comment by xemul** [comment]
OK, so there are two bugs here

1. if view_builder::consumer consumes end-of-stream _without_ consuming new partition it live-locks printing `view - Completed build step for base ks.tm_table, at token minimum token; views={}` 
2. with tablets, step 1 is triggered after tablets are cleaned up while building view for it

**Comment by xemul** [comment]
Looks like item 1 above is not a bug, but intent. There's a test_injected_noncritical_errors test in dtest that injects one-shot errors into builder and makes sure it eventually makes progress. Eventually -- once the problem (injection) disappears

**Comment by xemul** [comment]
So here's what happens.

First, builder starts building view-0 at starting token A. View-0 will be considered as built when it reaches token A again
It processes up to token B (B > A) when view-1 is added. Respectively, view-1's target token is B
It processes up to token C (C > B) when view-2 is added. And view-2's target token is C
More views are added eventually
It process up until end-of-stream of base table reader at token Z (Z > {C, B, A, ...})
Then reader is rewinded to start, i.e. to token A
At this point view-0 is built and marked as SUCCESS

Normally, the builder should process further and reach token B, thus marking view-1 as built. Then to C, marking view-2 as built and so on. Eventually it will mark all views as built and bail out.

But after reader is rewinded it does _not_ produce any more tokens, because tablets are migrated away from local node, so reading from it immediately produced end-of-stream, so reader is rewinded again and all not-yet-built views are attempted to be built again from base table beginning and this never happens.

**Comment by xemul** [comment]
Same thing should probably happen if the table truncated while view is being built, but it crashes (#17543) :shrug: even on vnodes case

**Comment by xemul** [comment]
Yup:
```
TRACE  ... [shard 0:strm] incremental_reader_selector - 0x6000057d6e00: created for range: [{minimum token, start},{maximum token, end}] with 0 sstables
TRACE  ... [shard 0:strm] incremental_reader_selector - 0x6000057d6e00: create_new_readers(null)
TRACE  ... [shard 0:strm] incremental_reader_selector - 0x6000057d6e00: 0 sstables to consider, advancing selector to {maximum token, w=1}
TRACE  ... [shard 0:strm] incremental_reader_selector - 0x6000057d6e00: created 0 new readers
```
